webapp/src/views/projects/languages/MachineTranslation/MachineTranslation.tsx

```diff
@@ -0,0 +1,317 @@
+  > = {};
+
+  languages.data?._embedded?.languages?.forEach((lang) => {
+    const config = settings.data?._embedded?.languageConfigs?.find(
+      (langSettings) => langSettings.targetLanguageId === lang.id
+    );
+
+    langDefaults[lang.tag] = {
+      enabled: config?.enabledServices || baseSetting?.enabledServices || [],
+      primary: config ? config.primaryService || 'none' : 'default',
+    };
+  });
+
+  const initialValues = {
+    default: {
+      enabled: baseSetting?.enabledServices,
+      primary: baseSetting?.primaryService || 'none',
+    },
+    languages: langDefaults,
+  };
+
+  const isFetching =
+    settings.isFetching || languages.isFetching || creditBalance.isFetching;
+
+  useGlobalLoading(isFetching);
+
+  const isUpdating = updateSettings.isLoading;
+
+  const formatLangSettings = (
+    lang: string | null,
+    enabled: string[],
+    primary: string | null | undefined
+  ): MachineTranslationLanguagePropsDto | null => {
+    const targetLanguageId = lang
+      ? languages.data?._embedded?.languages?.find((l) => l.tag === lang)?.id
+      : null;
+
+    if (primary === 'default') {
+      return null;
+    }
+
+    return {
+      targetLanguageId: targetLanguageId as any,
+      primaryService: (primary === 'none' ? null : primary) as any,
+      enabledServices: enabled as any,
+    };
+  };
+
+  useEffect(() => {
+    formRef.current?.resetForm();
+  }, [settings.data]);
+
+  useEffect(() => {
+    if (Number(settings.data?._embedded?.languageConfigs?.length) > 1) {
+      setExpanded(true);
+    }
+  }, [settings.data]);
+
+  return (
+    <>
+      {settings.data && languages.data && creditBalance.data && (
+        <div className={classes.container}>
+          <div className={tableClasses.table} style={{ gridTemplateColumns }}>
+            <Formik
+              key={formInstance}
+              initialValues={initialValues}
+              enableReinitialize={true}
+              onSubmit={() => {}}
+              validate={(values) => {
+                const languageSettings = Object.entries(values.languages).map(
+                  ([lang, value]) =>
+                    formatLangSettings(lang, value.enabled, value.primary)
+                );
+
+                languageSettings.push(
+                  formatLangSettings(
+                    null,
+                    values.default.enabled!,
+                    values.default.primary
+                  )
+                );
+
+                applyUpdate({
+                  settings: languageSettings.filter(
+                    Boolean
+                  ) as MachineTranslationLanguagePropsDto[],
+                });
+              }}
+              validateOnChange
+            >
+              {(form) => {
+                formRef.current = form;
+                return (
+                  <>
+                    <div className={tableClasses.topRow} />
+                    {providers.map((provider) => (
+                      <div
```

Comment:
Maybe create separate component for the provider? The file is a bit long...

Reply:
I`ll separate whole form content